### PR TITLE
Add linting via pre-commit

### PR DESCRIPTION
**Pull request by stefanv**
Add some standard formatting checks via pre-commit.

Eventually, it would be nice to add black or something similar here, but that'd need to be discussed first.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+  - repo: https://github.com/pre-commit/mirrors-prettier
+    rev: v2.7.1
+    hooks:
+      - id: prettier
+        files: \.(md|rst|yml|yaml)
+        args: [--prose-wrap=preserve]
+
+  - repo: https://github.com/asottile/pyupgrade
+    rev: v2.38.2
```
Comment (jarrodmillman) [inline .pre-commit-config.yaml:26]:
```suggestion
    rev: v3.0.0
```

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        name: isort (python)
+      - id: isort
+        name: isort (cython)
+        types: [cython]
+      - id: isort
+        name: isort (pyi)
+        types: [pyi]
+
+  - repo: https://gitlab.com/pycqa/flake8
+    rev: 3.8.4
```
Comment (jarrodmillman) [inline .pre-commit-config.yaml:44]:
```suggestion
    rev: 5.0.4
```

**Review by jarrodmillman — CHANGES_REQUESTED** [review]
LGTM. You can autolint in a separate PR, so that the commit hash can be added to `.git-blame-ignore-revs`.

**Comment by stefanv** [comment]
@jarrodmillman Do you think it's best to change files gradually as they get touched, or to do this as a one-off sweep as you suggest above?

**Comment by jarrodmillman** [comment]
I would do it once. I've done it in several projects and didn't run into any issues. Why not try it and if it causes issues just revert the change (or even git reset to the previous commit and git push to main)?

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+      - id: isort
+        name: isort (python)
+      - id: isort
+        name: isort (cython)
+        types: [cython]
+      - id: isort
+        name: isort (pyi)
+        types: [pyi]
+
+  - repo: https://gitlab.com/pycqa/flake8
```
Comment (lagru) [inline .pre-commit-config.yaml:43]: Wondering why you use the GitLab mirror here?

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -0,0 +1,48 @@
+# Install pre-commit hooks via
+# pre-commit install
+
+repos:
+  - repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: v4.3.0
```
Comment (lagru) [inline .pre-commit-config.yaml:6]: I think when using hooks from third parties, the same logic as for the GitHub actions applies and pinning to a specific commit is good practice. If you agree I'll do the work and suggest these changes. ;)

**Review by lagru — COMMENTED** [review]
Thanks for getting this started @stefanv. I'm not clear on how this will impact our workflow as I don't have much experience with this yet. Will we be able to override these hooks in case they prove to restrictive or have a bug?

**Comment by stefanv** [comment]
Go ahead and make the fixes you suggest, @lagru, they look fine to me.

The pre-commit is run in CI to ensure that everything conforms to the standards we define. Rewriting the code isn't necessary, but it also doesn't matter, so not sure it's worth figuring out how to switch it off.

**Comment by lagru** [comment]
Hmm, currently [pre-commit/action](https://github.com/pre-commit/action) is run on every file in our repo, which would preclude us from migrating step by step. There is https://github.com/pre-commit/action/issues/7 but that introduces another action dependency. 

I don't like the idea of asking a contributor to address all linting changes, e.g as found by flake8, in a long file himself even though only a small bit was changed. So I'd also advocate doing it once, though that PR might be a large barrier.

**Comment by jarrodmillman** [comment]
See #6568.

**Comment by stefanv** [comment]
We'll make a single commit with all linting changes, and add it to the `.git-blame-ignore-revs` to keep git blame history clean.

**Review by jarrodmillman — APPROVED** [review]

**Comment by stefanv** [comment]
I'm going to merge this to ensure that we don't get any other PRs merged on top of the linting patch.

File: .github/workflows/lint.yml [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+on: [push, pull_request]
+
+jobs:
+  pre-commit:
+    name: Lint
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - uses: actions/setup-python@v4
+      - uses: pre-commit/action@v3.0.0
```
[the remark this answers is not in the document]
Comment (jarrodmillman) [reply .github/workflows/lint.yml]: Let's deal with it later.

**Pull request merged**